specs/dynamicdependencies/DynamicDependencies.md

```diff
@@ -0,0 +1,1409 @@
+    /// all processes for that user is removed via PackageDependencyContext.Close()
+    /// (or process termination).
+    ///
+    /// Calls to Add() can be balanced by a PackageDependencyContext.Close()
+    /// to remove the entry from the package graph.
+    PackageDependencyContext Add();
+
+    /// Resolve a previously pinned PackageDependency to a specific package and
+    /// add it to the calling process' package graph. Once the dependency has
+    /// been added other code-loading methods (LoadLibrary, CoCreateInstance, etc)
+    /// can find the binaries in the resolved package.
+    ///
+    /// Package resoution is specific to a user. The same package dependency can
+    /// resolve to different packages for different users on a system.
+    ///
+    /// This adds the resulting package to the process' package graph.
+    /// A process' package graph is used to search for DLLs (per Dynamic-Link Library Search Order),
+    /// WinRT objects and other resources; the caller can now load DLLs, activate
+    /// WinRT objects and use other resources from the framework package until
+    /// PackageDependencyContext.Close() is called (or the process ends).
+    /// The package dependency Id must match a package dependency defined
+    /// for the calling user or the system (via PinForSystem) or an exception is raised.
+    ///
+    /// Each successful call adds the resolve packaged to the
+    /// calling process' package graph, even if already present. There is no
+    /// duplicate 'detection' or 'filtering' applied by the API (multiple
+    /// references to a package is not harmful). Once resolution is complete
+    /// the package stays resolved for that user until the last reference across
+    /// all processes for that user is removed via PackageDependencyContext.Close()
+    /// (or process termination).
+    ///
+    /// This adds the resolved package to the caller's package graph, per rank.
+    /// A process' package graph is a list of packages sorted by rank in ascending
+    /// order (-infinity...0...+infinity). If package(s) are present in the
+    /// package graph with the same rankrank the resolved package is
+    /// (by default) added after others of the same rank. To add a package
+    /// before others of the same rank, specify PackageDependency.PrependIfRankCollision.
+    ///
+    /// Calls to Add() can be balanced by a PackageDependencyContext.Close() (or object destruction)
+    /// to remove the entry from the package graph.
+    PackageDependencyContext Add(AddPackageDependencyOptions options);
+}
+
+/// A unique identifier for a resolved package dependency
+struct PackageDependencyContextId
+{
+    UInt64 Id;
+};
+
+/// This object provides access to information about a package dependency context
+/// and removes it from the caller's package graph when closed (via .Close() / .Dispose())
+/// or destroyed (i.e. undo packageDependency.Add()).
```

Comment:
So a WinRT app that wants to add a package to its graph for the lifetime of the process would just have to intentionally leak this object. That can be tricky in GC languages - you have to find something that escapes GC leak detection.

Reply:
DISCUSSION: The app would have to store this object. If this were not closable, we'd need to add a `.Remove` method instead.

RECOMMEND: Remove, `ICloseable` and then add a `Remove` method on the `Context` to manually drop the context.

CONSIDER: Ensure there's a sample showing stashing this into a global for long-lived use of the target FWP.